### PR TITLE
Test 2ministep

### PR DESCRIPTION
**Pull request by joakim-hove**
**Issue**
Part of #1069 

**Approach**
The purpose of this PR was to create a row_scaling test which demonstrates that two consecutive ministeps can be the same as one larger. However due to the internal random sampling it is difficult to get identical results - the test is therefor not that interesting.

The actual interesting test is in the last commit, in addition there are some small improvements/fixes in the preceeding commits

### CODE REVIEW

**Comment by ertomatic** [comment]
Can one of the admins verify this patch?

File: lib/enkf/local_config.cpp [hunk 1]
```diff
@@ -172,6 +172,9 @@ void local_config_free(local_config_type * local_config) {
 }
 
 local_ministep_type * local_config_alloc_ministep( local_config_type * local_config , const char * key, analysis_module_type* analysis_module) {
+  if (hash_has_key(local_config->ministep_storage, key))
```
Comment (ManInFez) [inline lib/enkf/local_config.cpp:175]: are we sure that this wont affect other users?

**Comment by ManInFez** [comment]
Test this please!

File: lib/enkf/local_config.cpp (hunk 1 above)
Comment (joakim-hove) [reply lib/enkf/local_config.cpp:175]: Well -  current master leads to a use-after-free situation; so this fix I am confident about.

**Comment by ManInFez** [comment]
#1126 is merged. Rebase this on top of master?

**Comment by joakim-hove** [comment]
Rebased!

**Comment by ManInFez** [comment]
Test this please!

File: python/res/enkf/data/field.py [hunk 2]
```diff
@@ -24,17 +24,28 @@ class Field(BaseCClass):
...
+        return self._get_size()
+
     def ijk_get_double(self, i, j, k):
         return self._ijk_get_double(i, j, k)
 
+    def __getitem__(self, index):
+        if index < len(self):
+            return self._iget_double(index)
+        else:
+            raise IndexError("Index: {} out of range: [0,{})", index, len(self))
```
Comment (ManInFez) [inline python/res/enkf/data/field.py]:
```suggestion
            raise IndexError(f"Index: {index} out of range: [0,{len(self)})")
```

File: lib/enkf/local_config.cpp (hunk 1 above)
Comment (ManInFez) [reply lib/enkf/local_config.cpp:175]: Very well, but it should either exit with a meaningful error or it should be dealt with here as well:
https://github.com/joakim-hove/libres/blob/388d0b29eb86a6745d5847f486c324ff0544dc39/lib/enkf/enkf_main.cpp#L1563

File: python/res/enkf/data/field.py (hunk 2 above)
Comment (joakim-hove) [reply python/res/enkf/data/field.py]: OK

File: lib/enkf/local_config.cpp (hunk 1 above)
Comment (joakim-hove) [reply lib/enkf/local_config.cpp:175]: I don't really understand what you request here. The python part - which is the user accessible will raise `KeyError()` - when it comes to the C++ function you link to that is called exactly once at object construction/initialization. I can of course add a 
```C++
if (!ministep) 
   throw std::logic_error("Failed to create ALL_ACTIVE ministep");
```
- or maybe I misunderstood your request?

Comment (ManInFez) [reply lib/enkf/local_config.cpp:175]: That is what I meant :) As we are changing the behavior of the function, it should be reflected wherever the function is used.

File: python/res/enkf/data/field.py [hunk 3]
```diff
@@ -24,17 +24,28 @@ class Field(BaseCClass):
...
         raise NotImplementedError("Class can not be instantiated directly!")
 
+    def __len__(self):
+        return self._get_size()
+
     def ijk_get_double(self, i, j, k):
         return self._ijk_get_double(i, j, k)
 
+    def __getitem__(self, index):
+        if index < len(self):
```
Comment (jondequinor) [inline python/res/enkf/data/field.py]:
```suggestion
        if index >= 0 and index < len(self):
```

File: tests/res/enkf/test_row_scaling_case.py [hunk 4]
```diff
@@ -370,3 +383,82 @@ def test_row_scaling_using_assign_vector(self):
...
+                field2 = node2.asField()
+                for iv, v in zip(init_field, field2):
+                    assert iv != v
+
+                    # If it were not for the random sampling in the
+                    # obs_data_allocE() function called once per ministep in
+                    # enkf_main_analysis_update() - we should have had this
+                    # assert:
+                    #
+                    # assert v1 == v2
```
Comment (jondequinor) [inline tests/res/enkf/test_row_scaling_case.py]:
```suggestion
```

Maybe add it to the top comment as a NOTE

File: python/res/enkf/data/field.py (hunk 3 above)
Comment (joakim-hove) [reply python/res/enkf/data/field.py]: ok

File: lib/enkf/local_config.cpp (hunk 1 above)
Comment (joakim-hove) [reply lib/enkf/local_config.cpp:175]: ok

**Comment by jondequinor** [comment]
> The purpose of this PR was to create a row_scaling test which demonstrates that two consecutive ministeps can be the same as one larger.

What does mean? The same as one larger? How is this connected to the diff?

File: tests/res/enkf/test_row_scaling_case.py (hunk 4 above)
Comment (joakim-hove) [reply tests/res/enkf/test_row_scaling_case.py]: Essentially the same information is already present at the top - I have just removed it.

File: lib/enkf/enkf_main.cpp [hunk 5]
```diff
@@ -1560,9 +1560,11 @@ void enkf_main_create_all_active_config( const enkf_main_type * enkf_main) {
   local_config_clear( local_config );
   {
     local_updatestep_type * default_step = local_config_get_updatestep(local_config);
-    local_ministep_type * ministep = local_config_alloc_ministep( local_config , "ALL_ACTIVE", NULL);
     local_obsdata_type * obsdata = local_config_alloc_obsdata(local_config, "ALL_OBS");
     local_dataset_type * all_active_dataset = local_config_alloc_dataset(local_config, "ALL_DATA");
+    local_ministep_type * ministep = local_config_alloc_ministep( local_config , "ALL_ACTIVE", NULL);
+    if (!ministep)
+      throw std::logic_error("Failed to create initial ALL_ACTIVE ministep")
```
Comment (ManInFez) [inline lib/enkf/enkf_main.cpp]:
```suggestion
      throw std::logic_error("Failed to create initial ALL_ACTIVE ministep");
```

Comment (joakim-hove) [reply lib/enkf/enkf_main.cpp]: Yes

**Comment by ManInFez** [comment]
Test this please!

**Comment by joakim-hove** [comment]
>>The purpose of this PR was to create a row_scaling test which demonstrates that two consecutive ministeps can be the same as one larger.

> What does mean? The same as one larger? How is this connected to the diff?

Well; 

I was asked by user whether it was possible/convenient to split an update with row scaling over two ministeps - so in order to demonstrate (and verify) this possibility I created this PR. The main content of this PR is the last commit which adds the test method:
```Python
def test_2ministep(self):
      ....
      ....
```
This test uses the row_scaling functionality, the first ministep scales layer 0 and the second ministep scales layer 1. Since the  updates involved are linear this should give identical results as one ministep update with scaling to both layers[1]. Due to stochastic sampling internally we unfortunately do not get identical results, but they are compared approximately.

[1]: Observe that the situation scaling to both layers is not explicit in the test code. The first update is just the defauilt update everything without any scaling case, and the scaling chosen for update two is 0 / 1 to ensure that the net effect is the same.

**Comment by ManInFez** [comment]
Test this please!

**Review by ManInFez — APPROVED** [review]

**Comment by jondequinor** [comment]
Thanks for explaining @joakim-hove!

**Pull request merged**